sns/parity_math.c

```diff
@@ -385,85 +768,911 @@ static void xor_calculate(struct m0_parity_math *math,
+			ret = M0_ERR_INFO(ret, "Failed to copy data from "
+					  "si_blocks[%u].sib_addr to "
+					  "out_bufs[%u].",
+					  rs->rs_failed_idx[i], i);
+			goto exit;
+		}
+	}
+
+	for (i = 0; i < ir->si_alive_nr; i++) {
+		if(rs->rs_alive_idx[i] == alive_block->sib_idx) {
+			curr_idx = i;
+			break;
+		}
+	}
+
+	if (curr_idx == UINT8_MAX) {
+		ret = M0_ERR_INFO(-EINVAL, "Failed to find alive block "
+				  "index %d in alive index array",
+				   alive_block->sib_idx);
+		goto exit;
+	}
+
+	/* Allocate buffer for input block. */
+	ret = m0_buf_alloc(&in_buf, length);
+	if (ret != 0) {
+		ret = BUF_ALLOC_ERR_INFO(ret, "IN", length);
+		goto exit;
+	}
+
+	/* Get data from current alive vector in input buffer. */
+	ret = m0_bufvec_to_buf_copy(&in_buf, alive_bufvec);
+	if (ret != 0) {
+		ret = M0_ERR_INFO(ret, "Failed to copy data from "
+				  "alive_bufvec to source buffer.");
+		goto exit;
+	}
+
+	/* Recover the data using input buffer and its index. */
+	isal_encode_data_update(out_bufs, &in_buf, curr_idx,
+				rs->rs_decode_tbls, ir->si_data_nr,
+				rs->rs_failed_nr);
+
+	/* Copy recovered data back to failed vectors. */
+	for (i = 0; i < rs->rs_failed_nr; i++) {
+		failed_bufvec = ir->si_blocks[rs->rs_failed_idx[i]].sib_addr;
+		ret = m0_buf_to_bufvec_copy(failed_bufvec, &out_bufs[i]);
+		if (ret != 0){
+			ret = M0_ERR_INFO(ret, "Failed to copy data from "
+					  "out_bufs[%u] to "
+					  "si_blocks[%u].sib_addr.", i,
+					  rs->rs_failed_idx[i]);
+			goto exit;
+		}
+	}
+
+exit:
+	m0_buf_free(&in_buf);
+	for (i = 0; i < rs->rs_failed_nr; i++)
+		m0_buf_free(&out_bufs[i]);
+
+	return M0_RC(ret);
+}
+
+static void isal_encode_data_update(struct m0_buf *dest_bufs, struct m0_buf *src_buf,
+				    uint32_t vec_idx, uint8_t *g_tbls,
+				    uint32_t data_nr, uint32_t dest_nr)
+{
+	uint32_t i;
+	uint32_t block_size;
+
+	M0_ENTRY("dest_bufs=%p, src_buf=%p, vec_idx=%u, "
+		 "g_tbls=%p, data_nr=%u, dest_nr=%u",
+		 dest_bufs, src_buf, vec_idx, g_tbls, data_nr, dest_nr);
+
+	M0_PRE(dest_bufs != NULL);
+	M0_PRE(src_buf != NULL);
+	M0_PRE(g_tbls != NULL);
+
+	uint8_t *dest_frags[dest_nr];
+
+	block_size = (uint32_t)src_buf->b_nob;
+
+	for (i = 0; i < dest_nr; ++i) {
+		BLOCK_SIZE_ASSERT_INFO(block_size, i, dest_bufs);
+		dest_frags[i] = (uint8_t *)dest_bufs[i].b_addr;
+	}
+
+	ec_encode_data_update(block_size, data_nr, dest_nr, vec_idx,
+			      g_tbls, (uint8_t *)src_buf->b_addr, dest_frags);
+
+	M0_LEAVE();
+}
+
+static void fails_sort(struct m0_reed_solomon *rs, uint8_t *fail,
+		       uint32_t total_count, uint32_t parity_count)
+{
+	uint32_t  i;
+	uint32_t  alive_nr;
+
+	M0_ENTRY();
+
+	M0_PRE(fail != NULL);
+	M0_PRE(rs != NULL);
+	M0_PRE(rs->rs_failed_idx != NULL);
+	M0_PRE(rs->rs_alive_idx != NULL);
+
+	rs->rs_failed_nr = 0;
+	for (i = 0, alive_nr = 0; i < total_count; i++) {
+		if (fail[i] != 0) {
+			VALUE_ASSERT_INFO(rs->rs_failed_nr < parity_count,
+					  rs->rs_failed_nr);
+			rs->rs_failed_idx[rs->rs_failed_nr++] = i;
+		}
+		else
```

Comment:
[Style] Move else to the same line as of closing braces.

Reply:
Done.